nilearn/connectome/group_sparse_cov.py

```diff
@@ -319,53 +307,32 @@ def _group_sparse_covariance(
     alpha2 = alpha**2
 
     for n in range(max_iter):
-        if max_norm is not None:
-            suffix = f" variation (max norm): {max_norm:.3e} "
-        else:
-            suffix = ""
-        if verbose > 1:
-            logger.log(
-                f"* iteration {n:d} "
-                f"({100.0 * n / max_iter:.0f} %){suffix} ...",
-                verbose=verbose,
-                stack_level=2,
-            )
+        suffix = (
+            f" variation (max norm): {max_norm:.3e} "
+            if max_norm is not None
+            else ""
+        )
+
+        logger.log(
+            f"* iteration {n:d} "
+            f"({100.0 * n / max_iter:.0f} %){suffix} ...",
+            verbose=verbose,
+            stack_level=2,
+        )
 
         omega_old[...] = omega
         for p in range(n_features):
             if p == 0:
-                # Initial state: remove first col/row
-                W = omega[1:, 1:, :].copy()  # stack of W(k)
-                W_inv = np.ndarray(shape=W.shape, dtype=np.float64)
-                for k in range(W.shape[2]):
-                    # stack of W^-1(k)
-                    W_inv[..., k] = scipy.linalg.inv(W[..., k])
-                    if debug:
-                        np.testing.assert_almost_equal(
-                            np.dot(W_inv[..., k], W[..., k]),
-                            np.eye(W_inv[..., k].shape[0]),
-                            decimal=10,
-                        )
-                        _assert_submatrix(omega[..., k], W[..., k], p)
-                        assert is_spd(W_inv[..., k])
+                W, W_inv = _set_initial_state_w_and_w_inv(omega, debug, p)
+
             else:
-                # Update W and W_inv
                 if debug:
                     omega_orig = omega.copy()
 
-                for k in range(n_subjects):
-                    _update_submatrix(
-                        omega[..., k], W[..., k], W_inv[..., k], p, h, v
-                    )
+                _update_w_and_w_inv(
+                    omega, debug, W, W_inv, n_subjects, p, h, v
+                )
 
-                    if debug:
-                        _assert_submatrix(omega[..., k], W[..., k], p)
-                        assert is_spd(W_inv[..., k], decimal=14)
-                        np.testing.assert_almost_equal(
-                            np.dot(W[..., k], W_inv[..., k]),
-                            np.eye(W_inv[..., k].shape[0]),
-                            decimal=10,
-                        )
                 if debug:
```

Comment:
there are a lot of those debug statement that in my opinion should not be there

once more of this code has been refactored it may be easier to replace them with proper tests

Reply:
Indeed.